runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -406,11 +407,14 @@ public void split(double fractionOfRemainder) {
 
         @Override
         public void close() throws Exception {
-          bundle.close();
-          currentClient.wrappedClient.unref();
-          if (loadBalanceBundles) {
-            availableCaches.offer(currentCache);
-            availableCachesSemaphore.release();
+          try {
+            bundle.close();
+          } finally {
+            currentClient.wrappedClient.unref();
+            if (loadBalanceBundles) {
+              availableCaches.offer(currentCache);
+              availableCachesSemaphore.release();
+            }
```

Comment:
Looks like this was responsible for the cleanup failing. bundle.close() may throw leaving the environment still referenced. My tests do not yield any more errors like in the description.